<a href="https://colab.research.google.com/github/luferIPCA/MIA-MLA-24-25/blob/main/1_DataSets_Essential.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Masters' in Applied Artificial Intelligence
## Machine Learning Algorithms Course

Notebooks for MLA course

by [*lufer*](mailto:lufer@ipca.pt)

(ver 2.0)

---

# Part I - Datasets for ML Modelling



*   This part explores the main considerations about datasets, namely: i) the way to get it; ii) main statistics involved; iii) essencial manipulations actions.
*   Main python topics and concepts are also explored.


**Contents**:

0. **Environment preparation**
1.   **Python essencial**
2.   **Understanding Datasets**
6.   **References**




## 0 - Environment preparation


Verify if python is installed; check the drive for notebooks files;

In [ ]:
#install phyton
#check if python is installed
!python --version # Python >= 3.9.18?

Python 3.10.12


### Import necessary Libraries

In [ ]:
import pandas as pd
import numpy as np

### Mounting Google Drive

In [ ]:
#mount my google onedrive
#from google.colab import drive
#drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
# or
#see https://towardsdatascience.com/different-ways-to-connect-google-drive-to-a-google-colab-notebook-pt-1-de03433d2f7a
#see https://adriandolinay.medium.com/an-introduction-to-google-colab-2023-6c26792827b3

from google.colab import drive

# it will ask for your google drive credentiaals
drive.mount('/content/gDrive/', force_remount=True)

Mounted at /content/gDrive/


### Unmounting Google Drive

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
#see https://www.geeksforgeeks.org/unmount-drive-in-google-colab-and-remount-to-another-drive/
#!ls "/content/MyDrive/"
#from google.colab import drive
#drive.flush_and_unmount()

## 1 - Python Essential


### Python Functions Definition

In [ ]:
%pwd


'/content'

In [ ]:
#phyton functions definition
def show(p): print (p)

#using defined fucntions
show("Lets start working with DataSets")

Lets start working with DataSets


Create an external libary of functions

In [ ]:
#Importing external python scripts (models)
%cd "/content/gdrive/MyDrive/Colab Notebooks/MIA - ML - 2024-2025/PythonFunctions/"
%pwd
#import
import MyFunctions as my  #Myfunction.py library
print(my.addNumbers(2,3))
print(my.getCurrentTime())



/content/gdrive/MyDrive/Colab Notebooks/MIA - ML - 2024-2025/PythonFunctions
5
2024-12-17 10:15:00


In [ ]:
#using module "os"
import os
print (os.getcwd())
os.listdir()
#os.rename()
#explore more functions
#see https://www.w3schools.com/python/module_os.asp

/content/gdrive/MyDrive/Colab Notebooks/MIA - ML - 2024-2025/PythonFunctions


['MyFunctions.py', '__pycache__']

In [ ]:
#using module shutil for copying, renaming, etc.
import shutil
shutil.copy("MyFunctions.py", "OtherCopied.py")
shutil.move("OtherCopied.py","OtherCopiedII.py")

'OtherCopiedII.py'

In [ ]:
os.listdir()

['MyFunctions.py', '__pycache__', 'OtherCopiedII.py']

In [ ]:
#Other example
import time #perf_counter(), time(), etc...see https://builtin.com/articles/timing-functions-python
print("Sleeping")
time.sleep(10) # sleep for a while; interrupt me!
#equivalent to
#for i in range(1,10):
#  time.sleep(1)
#  print(i)
print("Done Sleeping")



In [ ]:
#using time to know the time elapsed on the execution...
ini=time.perf_counter()
my.getCurrentTime()
fim=time.perf_counter()
a = f'{(fim-ini):.3}'
print("Elapsed time: ",a,"s")

Elapsed time:  0.00018 s


###More Python

In [ ]:
#Lists
l = [1,2,3,2,5,6]
#type(l)
len(l)
l[0]
l[-2]
l[-6]
#l[start_slice;value: end_slide:value]
l[2:4]
l[:]
l[2:]
l[:4]


list

In [ ]:
#Series: like an array
s = pd.Series([13,23,4,123,-4,134])
#type(s)
s

0     13
1     23
2      4
3    123
4     -4
5    134
dtype: int64

In [ ]:
#what happens?
#s = pd.Series([13,23,4,123],index=['a','b','c','d'])
#s
s[0]
s[1:3]
s[:]
s[:-1]
s[1:-1]

1     23
2      4
3    123
4     -4
dtype: int64

In [ ]:
#HashTable or Dictionary
# { label: serie,...,label:serie}

data = {
    "Colors":["Red","White","Orange","Black"],
    "Paintings":["Monalisa","The Kiss", "The Last Supper","O GLorioso"]
     }
type(data)
data["Colors"]
data["Paintings"]

['Monalisa', 'The Kiss', 'The Last Supper', 'O GLorioso']

In [ ]:
#Dataframes: two-dimensional array of series

df = pd.DataFrame(data)
type(df)
df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4 entries, 0 to 3
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   Colors     4 non-null      object
 1   Paintings  4 non-null      object
dtypes: object(2)
memory usage: 192.0+ bytes


array([['Red', 'Monalisa'],
       ['White', 'The Kiss'],
       ['Orange', 'The Last Supper'],
       ['Black', 'O GLorioso']], dtype=object)

### Python Classes and Objects

In [ ]:
#Class definition
#lufer
class Person:
  name=""
  age=1
  #-----
  # assign values to object properties, or other operations that are necessary
  # to do when the object is being created
  def __init__(self, name, age):
    #print("Inside __init__")
    self.name = name
    self.age = age
  #-----
  #  controls what should be returned when the class object is represented as a string
  def __str__(self):
    return f"Name: {self.name} (Age: {self.age})"
  #-----
  # other methods
  def myName(self):
    print("Hello my name is  " + self.name)

#Objects
p1 = Person("Ana Gustavo","12")   #uses __init__

print(p1)                         #uses __str__
print(p1.age)
print (f"Age: ({p1.age}) - Name: ({p1.name}) " )
p1.myName();
p1.name

Name: Ana Gustavo (Age: 12)
12
Age: (12) - Name: (Ana Gustavo) 
Hello my name is  Ana Gustavo


'Ana Gustavo'

### Código HTML

In [ ]:
%%html
<marquee style='width: 30%; color: red;'><b>O Benfica é o maior!</b></marquee>

### HTML/SVG

In [ ]:
%%html
<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 450 400" width="200" height="200">
  <rect x="80" y="60" width="250" height="250" rx="20" style="fill:red; stroke:black; fill-opacity:0.7" />
  <rect x="180" y="110" width="250" height="250" rx="40" style="fill:blue; stroke:black; fill-opacity:0.5;" />
</svg>

## 2 - Understanding Datasets bases

Datasets are sets o data to be used on models training. A dataset can have Continous values (quantities); Categorical values (fixed list values) or Text values.

### Read a Dataset from a CSV file

In [ ]:
filePath="/content/gDrive/MyDrive/Colab Notebooks/MIA - ML - 2024-2025/"

In [ ]:
#option 1
with open(filePath+'Datasets/test.csv', 'r') as f:
    print(f.read())

In [ ]:
#option 2
dataset = pd.read_csv(filePath+'/Datasets/test.csv')
#dataset is a dataframe

In [ ]:
dataset.head()

id  battery_power  blue  clock_speed  dual_sim  fc  four_g  int_memory  \
0   1           1043     1          1.8         1  14       0           5   
1   2            841     1          0.5         1   4       1          61   
2   3           1807     1          2.8         0   1       0          27   
3   4           1546     0          0.5         1  18       1          25   
4   5           1434     0          1.4         0  11       1          49   

   m_dep  mobile_wt  ...  pc  px_height  px_width   ram  sc_h  sc_w  \
0    0.1        193  ...  16        226      1412  3476    12     7   
1    0.8        191  ...  12        746       857  3895     6     0   
2    0.9        186  ...   4       1270      1366  2396    17    10   
3    0.5         96  ...  20        295      1752  3893    10     0   
4    0.5        108  ...  18        749       810  1773    15     8   

   talk_time  three_g  touch_screen  wifi  
0          2        0             1     0  
1          7        1             0     0  
2         10        0             1     1  
3          7        1             1     0  
4          7        1             0     1  

[5 rows x 21 columns]

In [ ]:
#type(dataset)
dataset.dtypes

After read, the dataset is a Dataframe python type.

Pandas DataFrame is like Excel-like Data.

In [ ]:
#option 3
# the "r" means ignoring special chars if the file name
f = open(filePath+r"/Datasets/test.csv", "r")
#f is a stream handler

In [ ]:
type(f)

_io.TextIOWrapper

###Generate Datasets

* Python offers a simple and powerful way to generate data...*Faker* is one of the libraries;

* Read more about "Sintetic Datasets"


In [ ]:
pip install faker

In [ ]:
from faker import Faker
import datetime
import random

# Set the seed for reproducibility
random.seed(42)

# Generate random sales data
fake = Faker()
sales_data = []

for _ in range(1000):
    city = fake.city()
    state = fake.state()
    sale = round(random.uniform(300, 5000), 2)
    sale_date = fake.date_between(start_date='-1y', end_date='today')
    sales_data.append([city, state, sale, sale_date])

# Create a DataFrame
df = pd.DataFrame(sales_data, columns=['City', 'State', 'Sale Amount', 'Sale Date'])

# Display the first few rows of the dataset
print(df.head())

                     City     State  Sale Amount   Sale Date
0  South Alexanderborough  Missouri      3305.31  2024-03-04
1             Monicamouth  Delaware       417.55  2024-12-05
2             Colleenfort  Arkansas      1592.64  2024-11-28
3             Lawsonshire   Vermont      1349.09  2024-02-07
4            Odonnellstad  Michigan      3761.41  2024-01-08


In [ ]:
#or using a function

def generateSales(num_customers):
    customers = []
    for _ in range(num_customers):
        customer_id = fake.uuid4()
        name = fake.name()
        email = fake.email()
        customers.append({'CustomerID': customer_id, 'CustomerName': name, 'Email': email})
    return pd.DataFrame(customers)

df1 = generateSales(100)
df1.head()

CustomerID     CustomerName  \
0  438f9b0b-4da9-422e-833b-c8f926ce47a4   Emily Williams   
1  2561f6b9-6e91-45c2-bc81-6c3e22260a43  Sarah Robertson   
2  899cc7da-00bb-4520-86b1-a0b80975c031    William Allen   
3  c849e6d8-6549-4580-989c-add41d6ab79c      Walter Hale   
4  ab64874c-432a-45a0-afd3-966d2b42b723      Shawn Lewis   

                         Email  
0       amandaneal@example.org  
1    angelajohnson@example.org  
2        ericahill@example.com  
3  campbellbradley@example.org  
4    barreracheryl@example.com

###Save a Dataset to file

In [ ]:
# Save the dataset to a CSV file
df.to_csv(filePath+'/Datasets/sales_dataset.csv', index=False)

### Datasets and Dataframes



Working with datasets in ML involves many functions, like:

*len(), shape, axes, index, values, info(), head(), tail(), drop(), iloc()*

**Number of records in the dataframe**

In [ ]:
#number of records in the dataframe
len(dataset)

1000

**Dataframe Structure (rows x columns)**

In [ ]:
#Structure (rows x columns) of the dataframe
dataset.shape

(1000, 21)

**Dataframe Axis (0 and 1)**

In [ ]:
dataset.axes
#RangeIndex = lines
#Index = columns

[RangeIndex(start=0, stop=1000, step=1),
 Index(['id', 'battery_power', 'blue', 'clock_speed', 'dual_sim', 'fc',
        'four_g', 'int_memory', 'm_dep', 'mobile_wt', 'n_cores', 'pc',
        'px_height', 'px_width', 'ram', 'sc_h', 'sc_w', 'talk_time', 'three_g',
        'touch_screen', 'wifi'],
       dtype='object')]

In [ ]:
#Columns
dataset.axes[1]

Index(['id', 'battery_power', 'blue', 'clock_speed', 'dual_sim', 'fc',
       'four_g', 'int_memory', 'm_dep', 'mobile_wt', 'n_cores', 'pc',
       'px_height', 'px_width', 'ram', 'sc_h', 'sc_w', 'talk_time', 'three_g',
       'touch_screen', 'wifi'],
      dtype='object')

In [ ]:
#what does dataset.axes[0]?

RangeIndex(start=0, stop=1000, step=1)

In [ ]:
dataset.index

RangeIndex(start=0, stop=1000, step=1)

In [ ]:
dataset.values

array([[1.000e+00, 1.043e+03, 1.000e+00, ..., 0.000e+00, 1.000e+00,
        0.000e+00],
       [2.000e+00, 8.410e+02, 1.000e+00, ..., 1.000e+00, 0.000e+00,
        0.000e+00],
       [3.000e+00, 1.807e+03, 1.000e+00, ..., 0.000e+00, 1.000e+00,
        1.000e+00],
       ...,
       [9.980e+02, 1.185e+03, 0.000e+00, ..., 1.000e+00, 0.000e+00,
        0.000e+00],
       [9.990e+02, 1.533e+03, 1.000e+00, ..., 0.000e+00, 1.000e+00,
        0.000e+00],
       [1.000e+03, 1.270e+03, 1.000e+00, ..., 1.000e+00, 0.000e+00,
        1.000e+00]])

**First 10 records**

In [ ]:
#first 10 records
dataset.head()

id  battery_power  blue  clock_speed  dual_sim  fc  four_g  int_memory  \
0   1           1043     1          1.8         1  14       0           5   
1   2            841     1          0.5         1   4       1          61   
2   3           1807     1          2.8         0   1       0          27   
3   4           1546     0          0.5         1  18       1          25   
4   5           1434     0          1.4         0  11       1          49   

   m_dep  mobile_wt  ...  pc  px_height  px_width   ram  sc_h  sc_w  \
0    0.1        193  ...  16        226      1412  3476    12     7   
1    0.8        191  ...  12        746       857  3895     6     0   
2    0.9        186  ...   4       1270      1366  2396    17    10   
3    0.5         96  ...  20        295      1752  3893    10     0   
4    0.5        108  ...  18        749       810  1773    15     8   

   talk_time  three_g  touch_screen  wifi  
0          2        0             1     0  
1          7        1             0     0  
2         10        0             1     1  
3          7        1             1     0  
4          7        1             0     1  

[5 rows x 21 columns]

**Last 10 records**

In [ ]:
#last 10 records
dataset.tail()

id  battery_power  blue  clock_speed  dual_sim  fc  four_g  int_memory  \
995   996           1700     1          1.9         0   0       1          54   
996   997            609     0          1.8         1   0       0          13   
997   998           1185     0          1.4         0   1       1           8   
998   999           1533     1          0.5         1   0       0          50   
999  1000           1270     1          0.5         0   4       1          35   

     m_dep  mobile_wt  ...  pc  px_height  px_width   ram  sc_h  sc_w  \
995    0.5        170  ...  17        644       913  2121    14     8   
996    0.9        186  ...   2       1152      1632  1933     8     1   
997    0.5         80  ...  12        477       825  1223     5     0   
998    0.4        171  ...  12         38       832  2509    15    11   
999    0.1        140  ...  19        457       608  2828     9     2   

     talk_time  three_g  touch_screen  wifi  
995         15        1             1     0  
996         19        0             1     1  
997         14        1             0     0  
998          6        0             1     0  
999          3        1             0     1  

[5 rows x 21 columns]

**Details about each column of the dataset**

Count of *Data type* and *Non-Null* values *Texto em itálico*

In [ ]:
#Details about each column
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             1000 non-null   int64  
 1   battery_power  1000 non-null   int64  
 2   blue           1000 non-null   int64  
 3   clock_speed    1000 non-null   float64
 4   dual_sim       1000 non-null   int64  
 5   fc             1000 non-null   int64  
 6   four_g         1000 non-null   int64  
 7   int_memory     1000 non-null   int64  
 8   m_dep          1000 non-null   float64
 9   mobile_wt      1000 non-null   int64  
 10  n_cores        1000 non-null   int64  
 11  pc             1000 non-null   int64  
 12  px_height      1000 non-null   int64  
 13  px_width       1000 non-null   int64  
 14  ram            1000 non-null   int64  
 15  sc_h           1000 non-null   int64  
 16  sc_w           1000 non-null   int64  
 17  talk_time      1000 non-null   int64  
 18  three_g  

Indexing just the rows

In [ ]:
#first column
dataset.iloc[0]

1043

In [ ]:
#first line (line 0)
dataset.iloc[[0]]

id  battery_power  blue  clock_speed  dual_sim  fc  four_g  int_memory  \
0   1           1043     1          1.8         1  14       0           5   

   m_dep  mobile_wt  ...  pc  px_height  px_width   ram  sc_h  sc_w  \
0    0.1        193  ...  16        226      1412  3476    12     7   

   talk_time  three_g  touch_screen  wifi  
0          2        0             1     0  

[1 rows x 21 columns]

In [ ]:
dataset.iloc[[0, 1]]

id  battery_power  blue  clock_speed  dual_sim  fc  four_g  int_memory  \
0   1           1043     1          1.8         1  14       0           5   
1   2            841     1          0.5         1   4       1          61   

   m_dep  mobile_wt  ...  pc  px_height  px_width   ram  sc_h  sc_w  \
0    0.1        193  ...  16        226      1412  3476    12     7   
1    0.8        191  ...  12        746       857  3895     6     0   

   talk_time  three_g  touch_screen  wifi  
0          2        0             1     0  
1          7        1             0     0  

[2 rows x 21 columns]

With a slice object

In [ ]:
dataset.iloc[:3]

id  battery_power  blue  clock_speed  dual_sim  fc  four_g  int_memory  \
0   1           1043     1          1.8         1  14       0           5   
1   2            841     1          0.5         1   4       1          61   
2   3           1807     1          2.8         0   1       0          27   

   m_dep  mobile_wt  ...  pc  px_height  px_width   ram  sc_h  sc_w  \
0    0.1        193  ...  16        226      1412  3476    12     7   
1    0.8        191  ...  12        746       857  3895     6     0   
2    0.9        186  ...   4       1270      1366  2396    17    10   

   talk_time  three_g  touch_screen  wifi  
0          2        0             1     0  
1          7        1             0     0  
2         10        0             1     1  

[3 rows x 21 columns]

Indexing lines and columns (both indexes)

In [ ]:
dataset.iloc[0,1] #iloc[row,column]

1043

In [ ]:
#lines from 1 to 3 and first 3 columns
dataset.iloc[1:3, 0:3]

id  battery_power  blue
1   2            841     1
2   3           1807     1

In [ ]:
#all
dataset.iloc[:, :]

id  battery_power  blue  clock_speed  dual_sim  fc  four_g  int_memory  \
0       1           1043     1          1.8         1  14       0           5   
1       2            841     1          0.5         1   4       1          61   
2       3           1807     1          2.8         0   1       0          27   
3       4           1546     0          0.5         1  18       1          25   
4       5           1434     0          1.4         0  11       1          49   
..    ...            ...   ...          ...       ...  ..     ...         ...   
995   996           1700     1          1.9         0   0       1          54   
996   997            609     0          1.8         1   0       0          13   
997   998           1185     0          1.4         0   1       1           8   
998   999           1533     1          0.5         1   0       0          50   
999  1000           1270     1          0.5         0   4       1          35   

     m_dep  mobile_wt  ...  pc  px_height  px_width   ram  sc_h  sc_w  \
0      0.1        193  ...  16        226      1412  3476    12     7   
1      0.8        191  ...  12        746       857  3895     6     0   
2      0.9        186  ...   4       1270      1366  2396    17    10   
3      0.5         96  ...  20        295      1752  3893    10     0   
4      0.5        108  ...  18        749       810  1773    15     8   
..     ...        ...  ...  ..        ...       ...   ...   ...   ...   
995    0.5        170  ...  17        644       913  2121    14     8   
996    0.9        186  ...   2       1152      1632  1933     8     1   
997    0.5         80  ...  12        477       825  1223     5     0   
998    0.4        171  ...  12         38       832  2509    15    11   
999    0.1        140  ...  19        457       608  2828     9     2   

     talk_time  three_g  touch_screen  wifi  
0            2        0             1     0  
1            7        1             0     0  
2           10        0             1     1  
3            7        1             1     0  
4            7        1             0     1  
..         ...      ...           ...   ...  
995         15        1             1     0  
996         19        0             1     1  
997         14        1             0     0  
998          6        0             1     0  
999          3        1             0     1  

[1000 rows x 21 columns]

Droping data

In [ ]:
dataset.drop(['blue', 'four_g'], axis=1)
#or
#dataset.drop(columns=['blue, 'four_g'])
#or
#dataset.drop(['blue', 'four_g'], axis='columns')

In [ ]:
pd.set_option("display.precision", 2)

### Basic Statistics over a Dataset

Basic statistics with 'describe()'

1. For Numeric Columns (default behavior):

* It provides summary statistics such as:
 * Count: The number of non-missing values.
 * Mean: The average of the values.
 * Std: The standard deviation (measure of spread).
 * Min: The smallest value.
 * 25%: The first quartile (25th percentile).
 * 50%: The median (50th percentile).
 * 75%: The third quartile (75th percentile).
 * Max: The largest value.
2.  For Non-Numeric Columns (if specified using include='all' or include=object):

* It provides:
 * Count: The number of non-missing values.
 * Unique: The number of unique values.
 * Top: The most frequent value.
 * Freq: The frequency of the top value.

3. **Key Notes about 'describe()'**
* Non-numeric data is ignored unless explicitly included using include.
* The method helps quickly identify:
 * Missing data (count vs. total rows).
 * Outliers (e.g., large differences between min, max, and mean).
 * Skewness (if the mean and median differ significantly).

 * **Default Percentiles**: percentiles=None computes the 25th,  50th, and 75th percentiles by default.
* **Default Inclusion**: exclude=None includes numeric columns only, unless explicitly overridden.
* Explicit Parameters:
 * Use percentiles to customize which percentiles to compute.
 * Use include='all' to analyze both numeric and non-numeric columns.
 * Use exclude to omit specific data types (e.g., exclude=[float]).

In [ ]:
#see https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.describe.html
#by default
dataset.describe()
#what happens with
#dataset.describe(include='all')
#How do you explain it?

id  battery_power         blue  clock_speed     dual_sim  \
count  1000.000000    1000.000000  1000.000000  1000.000000  1000.000000   
mean    500.500000    1248.510000     0.516000     1.540900     0.517000   
std     288.819436     432.458227     0.499994     0.829268     0.499961   
min       1.000000     500.000000     0.000000     0.500000     0.000000   
25%     250.750000     895.000000     0.000000     0.700000     0.000000   
50%     500.500000    1246.500000     1.000000     1.500000     1.000000   
75%     750.250000    1629.250000     1.000000     2.300000     1.000000   
max    1000.000000    1999.000000     1.000000     3.000000     1.000000   

                fc       four_g   int_memory        m_dep   mobile_wt  ...  \
count  1000.000000  1000.000000  1000.000000  1000.000000  1000.00000  ...   
mean      4.593000     0.487000    33.652000     0.517500   139.51100  ...   
std       4.463325     0.500081    18.128694     0.280861    34.85155  ...   
min       0.000000     0.000000     2.000000     0.100000    80.00000  ...   
25%       1.000000     0.000000    18.000000     0.300000   109.75000  ...   
50%       3.000000     0.000000    34.500000     0.500000   139.00000  ...   
75%       7.000000     1.000000    49.000000     0.800000   170.00000  ...   
max      19.000000     1.000000    64.000000     1.000000   200.00000  ...   

                pc    px_height     px_width          ram         sc_h  \
count  1000.000000  1000.000000  1000.000000  1000.000000  1000.000000   
mean     10.054000   627.121000  1239.774000  2138.998000    11.995000   
std       6.095099   432.929699   439.670981  1088.092278     4.320607   
min       0.000000     0.000000   501.000000   263.000000     5.000000   
25%       5.000000   263.750000   831.750000  1237.250000     8.000000   
50%      10.000000   564.500000  1250.000000  2153.500000    12.000000   
75%      16.000000   903.000000  1637.750000  3065.500000    16.000000   
max      20.000000  1907.000000  1998.000000  3989.000000    19.000000   

              sc_w    talk_time      three_g  touch_screen         wifi  
count  1000.000000  1000.000000  1000.000000    1000.00000  1000.000000  
mean      5.316000    11.085000     0.756000       0.50000     0.507000  
std       4.240062     5.497636     0.429708       0.50025     0.500201  
min       0.000000     2.000000     0.000000       0.00000     0.000000  
25%       2.000000     6.750000     1.000000       0.00000     0.000000  
50%       5.000000    11.000000     1.000000       0.50000     1.000000  
75%       8.000000    16.000000     1.000000       1.00000     1.000000  
max      18.000000    20.000000     1.000000       1.00000     1.000000  

[8 rows x 21 columns]

**DataFrame.describe(percentiles=None, include=None, exclude=None)**
* percentiles normal (25%, 59%, 75% (default)
* include only numeric values (default)
* exclude NaN values (default)


In [ ]:
#"None" is the default...means consider only numerical values
dataset.describe(percentiles=None, exclude=None)

Considering the percentiles of *battery_power* column:

*   (25%) - 25% of the records have the value 895 or below
*   (50%) - 50% of the records have the value 1246.5 or below
*   (75%) - 75% of your records have the value 1629.25 or below

In [ ]:
dataset.describe(include='all')



---


Practice: Explore the simple Lab

In [ ]:
import pandas as pd

data = {'A': [1, 2, 3, 4, 5], 'B': [np.nan, 20, 30, 40, 50], 'C': ['a', 'b', 'c', 'd', 'c']}
df = pd.DataFrame(data)

In [ ]:
df

A     B  C
0  1   NaN  a
1  2  20.0  b
2  3  30.0  c
3  4  40.0  d
4  5  50.0  c

In [ ]:
df.shape

(5, 3)

In [ ]:
#basic statistics for numerical only
print(df.describe())

              A          B
count  5.000000   4.000000
mean   3.000000  35.000000
std    1.581139  12.909944
min    1.000000  20.000000
25%    2.000000  27.500000
50%    3.000000  35.000000
75%    4.000000  42.500000
max    5.000000  50.000000


In [ ]:
df.describe(include='all')

A          B    C
count   5.000000   4.000000    5
unique       NaN        NaN    4
top          NaN        NaN    c
freq         NaN        NaN    2
mean    3.000000  35.000000  NaN
std     1.581139  12.909944  NaN
min     1.000000  20.000000  NaN
25%     2.000000  27.500000  NaN
50%     3.000000  35.000000  NaN
75%     4.000000  42.500000  NaN
max     5.000000  50.000000  NaN



---



In [ ]:
#What it does?
df.describe(percentiles=None, exclude=None)

#and this
df.describe(percentiles=[0.1, 0.9])

**Statistical calculus for a particular column**

*Count number of non-NA/null observations.*

In [ ]:
dataset['battery_power'].count()

1000

*Maximum of the battery_power values*

In [ ]:
dataset['battery_power'].max()

1999

*Mean of the battery_power values*

In [ ]:
dataset['battery_power'].mean()

1248.51

*Minimum value of battery_power*

In [ ]:
dataset['battery_power'].min()

500

*Median of battery_power column*

In [ ]:
dataset['battery_power'].median()

1246.5

*Check how may occurences for each value of a particulaar column*

In [ ]:
dataset['battery_power'].value_counts()

battery_power
1074    5
1981    5
529     4
1745    4
1715    4
       ..
1248    1
1392    1
1706    1
1297    1
1185    1
Name: count, Length: 721, dtype: int64

*Select part of the columns*

In [ ]:
dataset[['battery_power','dual_sim']]

battery_power  dual_sim
0             1043         1
1              841         1
2             1807         0
3             1546         1
4             1434         0
..             ...       ...
995           1700         0
996            609         1
997           1185         0
998           1533         1
999           1270         0

[1000 rows x 2 columns]

*Details about part of the columns*

In [ ]:
dataset[['battery_power','dual_sim']].describe(include="all")

battery_power     dual_sim
count    1000.000000  1000.000000
mean     1248.510000     0.517000
std       432.458227     0.499961
min       500.000000     0.000000
25%       895.000000     0.000000
50%      1246.500000     1.000000
75%      1629.250000     1.000000
max      1999.000000     1.000000

*Details about a partivular column*

In [ ]:
dataset.battery_power.describe()

count    1000.000000
mean     1248.510000
std       432.458227
min       500.000000
25%       895.000000
50%      1246.500000
75%      1629.250000
max      1999.000000
Name: battery_power, dtype: float64

### Considering columns of type *objet* (text)

By default, only numeric values are considered. To analyse objects columns (text values) use "*dataset.describe(include=object)*"

*Example of Dataframe with objects:*

In [ ]:
df = pd.DataFrame({'gender': pd.Categorical(['Male', 'Female', 'Others']),
                   'quantity': [12, 25, 3],
                   'symbol': ['M', 'F', 'O']
                   })
df

gender  quantity symbol
0    Male        12      M
1  Female        25      F
2  Others         3      O

*By default omly numerical values are described*

In [ ]:
df.describe()

quantity
count   3.000000
mean   13.333333
std    11.060440
min     3.000000
25%     7.500000
50%    12.000000
75%    18.500000
max    25.000000

*Describing all columns of a DataFrame regardless of data type*

In [ ]:
df.describe(include="all")

gender   quantity symbol
count        3   3.000000      3
unique       3        NaN      3
top     Female        NaN      M
freq         1        NaN      1
mean       NaN  13.333333    NaN
std        NaN  11.060440    NaN
min        NaN   3.000000    NaN
25%        NaN   7.500000    NaN
50%        NaN  12.000000    NaN
75%        NaN  18.500000    NaN
max        NaN  25.000000    NaN

*Including only string columns in the DataFrame description.*

In [ ]:
df.describe(include=object)
#or
df.describe(include=[object])

symbol
count       3
unique      3
top         M
freq        1

*Excluding numeric columns from a DataFrame description.*

In [ ]:
df.describe(exclude=[np.number])

gender symbol
count        3      3
unique       3      3
top     Female      M
freq         1      1

*Including only numeric columns in the DataFrame description.*

In [ ]:
df.describe(include=[np.number])

quantity
count   3.000000
mean   13.333333
std    11.060440
min     3.000000
25%     7.500000
50%    12.000000
75%    18.500000
max    25.000000

*Only Categorical columns*

In [ ]:

df.describe(include=['category'])

gender
count        3
unique       3
top     Female
freq         1

*Excluding object columns from a DataFrame description.*

In [ ]:
df.describe(exclude=[object])

gender   quantity
count        3   3.000000
unique       3        NaN
top     Female        NaN
freq         1        NaN
mean       NaN  13.333333
std        NaN  11.060440
min        NaN   3.000000
25%        NaN   7.500000
50%        NaN  12.000000
75%        NaN  18.500000
max        NaN  25.000000

### Filtering Data

The idea it to *get filtered data* for any criterium.

Column value greater than...

In [ ]:
#Back to original dataset
newdf = dataset.copy()
newdf = dataset.query('battery_power>1000')
newdf

id  battery_power  blue  clock_speed  dual_sim  fc  four_g  int_memory  \
0       1           1043     1          1.8         1  14       0           5   
2       3           1807     1          2.8         0   1       0          27   
3       4           1546     0          0.5         1  18       1          25   
4       5           1434     0          1.4         0  11       1          49   
5       6           1464     1          2.9         1   5       1          50   
..    ...            ...   ...          ...       ...  ..     ...         ...   
992   993           1895     0          0.5         1   0       1          62   
995   996           1700     1          1.9         0   0       1          54   
997   998           1185     0          1.4         0   1       1           8   
998   999           1533     1          0.5         1   0       0          50   
999  1000           1270     1          0.5         0   4       1          35   

     m_dep  mobile_wt  ...  pc  px_height  px_width   ram  sc_h  sc_w  \
0      0.1        193  ...  16        226      1412  3476    12     7   
2      0.9        186  ...   4       1270      1366  2396    17    10   
3      0.5         96  ...  20        295      1752  3893    10     0   
4      0.5        108  ...  18        749       810  1773    15     8   
5      0.8        198  ...   9        569       939  3506    10     7   
..     ...        ...  ...  ..        ...       ...   ...   ...   ...   
992    0.9         99  ...   0       1019      1698  2563    10     8   
995    0.5        170  ...  17        644       913  2121    14     8   
997    0.5         80  ...  12        477       825  1223     5     0   
998    0.4        171  ...  12         38       832  2509    15    11   
999    0.1        140  ...  19        457       608  2828     9     2   

     talk_time  three_g  touch_screen  wifi  
0            2        0             1     0  
2           10        0             1     1  
3            7        1             1     0  
4            7        1             0     1  
5            3        1             1     1  
..         ...      ...           ...   ...  
992         13        1             0     1  
995         15        1             1     0  
997         14        1             0     0  
998          6        0             1     0  
999          3        1             0     1  

[666 rows x 21 columns]

Filtering one column and compare with another

In [ ]:
newdf.loc[newdf["battery_power"] >1500, "clock_speed"].value_counts()

clock_speed
0.5    70
1.6    16
0.6    14
2.1    14
2.0    13
2.8    12
2.6    12
1.3    12
1.4    12
2.5    11
1.1    11
0.8    10
2.3    10
2.7    10
0.7     9
1.2     9
2.2     9
1.7     9
1.9     9
1.8     9
1.0     8
2.4     8
0.9     7
1.5     7
2.9     6
3.0     5
Name: count, dtype: int64

Complex filtering

In [ ]:
newdf[(newdf["clock_speed"] < 1) & (newdf["dual_sim"] ==1)].groupby(\

    ["battery_power", "mobile_wt","clock_speed"])["int_memory"].count()

battery_power  mobile_wt  clock_speed
1010           195        0.7            1
1018           152        0.6            1
1021           184        0.9            1
1031           82         0.5            1
1038           99         0.5            1
                                        ..
1979           136        0.5            1
1980           141        0.5            1
1982           98         0.8            1
1985           163        0.5            1
1999           183        0.5            1
Name: int_memory, Length: 109, dtype: int64

*Calculus over filtered data*

In [ ]:
dataset.groupby('dual_sim').sum()

id  battery_power  blue  clock_speed    fc  four_g  int_memory  \
dual_sim                                                                       
0         242134         616243   252        749.4  2090     229       16364   
1         258366         632267   264        791.5  2503     258       17288   

          m_dep  mobile_wt  n_cores    pc  px_height  px_width      ram  sc_h  \
dual_sim                                                                        
0         246.9      67414     2094  4631     301420    595383  1006957  5780   
1         270.6      72097     2234  5423     325701    644391  1132041  6215   

          sc_w  talk_time  three_g  touch_screen  wifi  
dual_sim                                                
0         2572       5342      365           233   237  
1         2744       5743      391           267   270

In [ ]:
dataset.groupby('dual_sim').count()

id  battery_power  blue  clock_speed   fc  four_g  int_memory  \
dual_sim                                                                   
0         483            483   483          483  483     483         483   
1         517            517   517          517  517     517         517   

          m_dep  mobile_wt  n_cores   pc  px_height  px_width  ram  sc_h  \
dual_sim                                                                   
0           483        483      483  483        483       483  483   483   
1           517        517      517  517        517       517  517   517   

          sc_w  talk_time  three_g  touch_screen  wifi  
dual_sim                                                
0          483        483      483           483   483  
1          517        517      517           517   517

In [ ]:
dataset.groupby("dual_sim", sort=False)["n_cores"].sum()

dual_sim
1    2234
0    2094
Name: n_cores, dtype: int64

## 3 - Final Pratice

In [ ]:
#explore the dataframe
data = {
    'A': ['apple', 'banana', 'apple', 'orange', 'banana'],
    'B': [10, 20, None, 40, 50],
    'C': ['x', 'y', 'x', 'z', None]
}
df = pd.DataFrame(data)

## References


*   Using pandas and Python to Explore Your Dataset
https://realpython.com/pandas-python-explore-dataset/




End!